### ISSUE TITLE
hope for support to indexing array of struct type 

### ISSUE THREAD

**Issue by KimDahye** [lucidworks/spark-solr#141]
Hello, I am Sophie.
My team is building a searching engine with Solr 6.5.0 and processing data with Spark 2.1.0. So we are considering spark-solr library to insert dataframes to Solr. We want to index data with `childDocuments` structure (referred from [this document](https://cwiki.apache.org/confluence/display/solr/Uploading+Data+with+Index+Handlers#UploadingDatawithIndexHandlers-NestedChildDocuments)) though, spark-solr dose not support this structure. Could you support this array of struct type structure as soon as possible? 

#### Test Codes
- sample.json
```json
{ "user":"gT35Hhhre9m", "dates":[ "2016-01-29",  "2016-01-28"], "status":"OK", "_childDocuments_":[{ "foo":123,"bar":"val1"},{ "foo":456,"bar":"val2"}]}
{ "user":"gT35Hhhre9n", "dates":[ "2016-01-29",  "2016-01-28"], "status":"OK", "_childDocuments_":[{ "foo":789,"bar":"val1"},{ "foo":1011,"bar":"val2"}]}
```

- scala code & output
```scala
val df = spark.read.json("sample.json")
df.printSchema
df.show()

val options = Map(
  "zkhost" -> "OUR_ZKHOST",
  "collection" -> "sample",
  "flatten_multivalued" -> "false"
)

df.write.format("solr").options(options).mode(org.apache.spark.sql.SaveMode.Overwrite).save
```

```console
df: org.apache.spark.sql.DataFrame = [_childDocuments_: array<struct<bar:string,foo:bigint>>, dates: array<string> ... 2 more fields]
root
 |-- _childDocuments_: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- bar: string (nullable = true)
 |    |    |-- foo: long (nullable = true)
 |-- dates: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- status: string (nullable = true)
 |-- user: string (nullable = true)
+--------------------+--------------------+------+-----------+
|    _childDocuments_|               dates|status|       user|
+--------------------+--------------------+------+-----------+
|[[val1,123], [val...|[2016-01-29, 2016...|    OK|gT35Hhhre9m|
|[[val1,789], [val...|[2016-01-29, 2016...|    OK|gT35Hhhre9n|
+--------------------+--------------------+------+-----------+
```
- solr output
We expect output as this:
```json
{
  "response":{"numFound":2,"start":0,"docs":[
      {
        "_childDocuments_":[
          {
            "bar" : "val1",
            "foo": 123,
          },
          {
            "bar" : "val2",
            "foo": 456,
          }
        ],
        "dates":["2016-01-29",
          "2016-01-28"],
        "status":"OK",
        "user":"gT35Hhhre9m",
        "id":"a1b79369-b34f-4022-b7ab-9ac93ba8c0fb",
        "_version_":1565897807074689024},
      {
        "_childDocuments_":[
        {
          "bar" : "val1",
          "foo": 789,
        },
        {
          "bar" : "val2",
          "foo": 1011,
        }
        ],
        "dates":["2016-01-29",
          "2016-01-28"],
        "status":"OK",
        "user":"gT35Hhhre9n",
        "id":"fadf0a44-03b7-4375-8140-949fec34d5b7",
        "_version_":1565897807077834752}]
  }
}
```

but it is like this:
```json
{
  "response":{"numFound":2,"start":0,"docs":[
      {
        "_childDocuments_":["org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema:[val1,123]",
          "org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema:[val2,456]"],
        "dates":["2016-01-29",
          "2016-01-28"],
        "status":"OK",
        "user":"gT35Hhhre9m",
        "id":"a1b79369-b34f-4022-b7ab-9ac93ba8c0fb",
        "_version_":1565897807074689024},
      {
        "_childDocuments_":["org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema:[val1,789]",
          "org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema:[val2,1011]"],
        "dates":["2016-01-29",
          "2016-01-28"],
        "status":"OK",
        "user":"gT35Hhhre9n",
        "id":"fadf0a44-03b7-4375-8140-949fec34d5b7",
        "_version_":1565897807077834752}]
  }
}
```

#### spark-solr `insert` function
As you know, in [this code](https://github.com/lucidworks/spark-solr/blob/982a4e9848d57b1f3998257cb84df394f22c6af4/src/main/scala/com/lucidworks/spark/SolrRelation.scala#L648), I found that there is no consideration about inserting array of struct type. Also, there is no consideration about adding fields of that type - [code](https://github.com/lucidworks/spark-solr/blob/982a4e9848d57b1f3998257cb84df394f22c6af4/src/main/scala/com/lucidworks/spark/SolrRelation.scala#L592).

**Comment by kiranchitturi** [comment]
@KimDahye 

spark-solr currently does not support indexing nested documents to Solr. In this case, we have to treat `_childDocuments_` as a special field and not try to define it and handle it's value (StructType) differently

We do not have any current plans on supporting this feature. However, feel free to work on this and open a PR

**Comment by KimDahye** [comment]
@kiranchitturi 
Thank you for answering. I should try :)

**Comment by kiranchitturi** [comment]
Fixed in #145
